p2p/ipld/read.go

```diff
@@ -224,25 +224,35 @@ func GetLeafData(
 	rootCid cid.Cid,
 	leafIndex uint32,
 	totalLeafs uint32, // this corresponds to the extended square width
-	api coreiface.CoreAPI,
+	dag format.NodeGetter,
 ) ([]byte, error) {
-	// calculate the path to the leaf
-	leafPath, err := leafPath(leafIndex, totalLeafs)
+	nd, err := GetTreeLeaf(ctx, dag, rootCid, leafIndex, totalLeafs)
 	if err != nil {
 		return nil, err
 	}
 
-	// use the root cid and the leafPath to create an ipld path
-	p := path.Join(path.IpldPath(rootCid), leafPath...)
+	return nd.RawData()[1:], nil
+}
 
-	// resolve the path
-	node, err := api.ResolveNode(ctx, p)
+func GetTreeLeaf(ctx context.Context, dag format.NodeGetter, root cid.Cid, leaf, total uint32) (format.Node, error) {
+	nd, err := dag.Get(ctx, root)
 	if err != nil {
 		return nil, err
 	}
 
-	// return the leaf, without the nmt-leaf-or-node byte
-	return node.RawData()[1:], nil
+	lnks := nd.Links()
+	if len(lnks) == 1 {
+		return dag.Get(ctx, lnks[0].Cid)
+	}
+
+	total /= 2
+	if leaf < total {
+		root = lnks[0].Cid
+	} else {
+		root, leaf = lnks[1].Cid, leaf-total
+	}
```

Comment:
mind adding some documentation somewhere? doesn't have to be anything too elaborate, but I think some simple guidance would make a big difference. Particularly for readers who are not familiar with our plugin.

Reply:
Sure. Also, that’s no the final form. I think to close this PR actually and rewrite it afresh to be based on the master, but that requires #323 to be merged first.

Reply:
merged :-)

Reply:
I added some documentation now, hope it will help understand the logic

Reply:
Cool! Just saw that after reviewing.
> Use of Dag Sessions(optimization)

I think I understand this now. Although it's not clear to me how it works under the hood. But it makes sense intuitively.